Google.Apis.Common/src/Google.Apis.Common/Paginator.cs

```diff
@@ -0,0 +1,152 @@
+        public IAsyncEnumerable<TResource> FetchAsync(TRequest initialRequest, Func<TRequest, CancellationToken, Task<TResponse>> responseFetcher)
+        {
+            Preconditions.CheckNotNull(responseFetcher, nameof(responseFetcher));
+            TRequest nextRequest = initialRequest;
+            bool done = false;
+            return new PagingAsyncEnumerable<TResource>(async cancellationToken =>
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                if (done)
+                {
+                    return null;
+                }
+                TResponse response = await responseFetcher(nextRequest, cancellationToken).ConfigureAwait(false);
+                TToken token = _tokenExtractor(response);
+                done = IsEmptyToken(token);
+                nextRequest = _requestProvider(initialRequest, token);
+                return _resourceExtractor(response);
+            });
+        }
+
+        /// <summary>
+        /// Eagerly (but asynchronously) fetches a complete set of resources, potentially making multiple requests.
+        /// </summary>
+        /// <param name="initialRequest">The initial request to send. If this contains a page token,
+        /// that token is maintained. This is also passed to the request provider along with a token
+        /// to provide another request. (The provider may build a new object, or modify the existing request.)</param>
+        /// <param name="responseFetcher">A function to asynchronously fetch a response containing a
+        /// page of results given a request. Must not be null.</param>
+        /// <returns>A sequence of resources, which are fetched asynchronously and a page at a time.</returns>
+        /// <param name="cancellationToken"></param>
+        /// <returns>A task </returns>
+        public async Task<IList<TResource>> FetchAllAsync(
+            TRequest initialRequest,
+            Func<TRequest, CancellationToken, Task<TResponse>> responseFetcher,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Preconditions.CheckNotNull(responseFetcher, nameof(responseFetcher));
+            List<TResource> results = new List<TResource>();
+            TToken token;
+            TRequest nextRequest = initialRequest;
+            do
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                TResponse response = await responseFetcher(nextRequest, cancellationToken).ConfigureAwait(false);
+                token = _tokenExtractor(response);
+                nextRequest = _requestProvider(initialRequest, token);
+                results.AddRange(_resourceExtractor(response));
+            } while (!IsEmptyToken(token));
+            return results;
+        }
+
+        private bool IsEmptyToken(TToken token)
+        {
+            return EqualityComparer<TToken>.Default.Equals(token, _emptyToken);
```

Comment:
Wondering about the case when the empty token is something like a `string`, where it could be either `null` or empty and both mean the same thing, should we worry about it and if so provide a callback to determine if the token is the _empty_ one instead of comparing the value ourselves?